tests/parameter_search/test_automatic_rounding.py

```diff
@@ -0,0 +1,240 @@
+
+    tlu_node = tlu_nodes[0]
+
+    # Verify that this TLU was optimized
+    if tlu_node in tlu_optimizer.statistics:
+        assert (
+            set(["shape", "size", "original_bitwidth", "optimized_bitwidth"])
+            == tlu_optimizer.statistics[tlu_node].keys()
+        )
+
+    scales_tlu = tlu_node.properties["attributes"]["opt_round_a"]
+    offsets_tlu = tlu_node.properties["attributes"]["opt_round_b"]
+
+    if tlu_test_mode == "per_tensor":
+        assert offsets_tlu.size == 1
+        assert scales_tlu.size == 1
+    elif tlu_test_mode == "per_cell":
+        assert offsets_tlu.size == model.input_shape[2] * model.input_shape[3]
+        assert scales_tlu.size == model.input_shape[2] * model.input_shape[3]
+    elif tlu_test_mode == "per_neuron":
+        assert offsets_tlu.size == model.n_neurons
+        assert scales_tlu.size == model.n_neurons
+    else:
+        raise AssertionError("Invalid tlu granularity test mode")
+
+
+@pytest.mark.parametrize("n_bits", range(2, 6))
+def test_tlu_analysis_optimization(load_data, n_bits):
+    """Test that the exact rounding parameters are found for NN with power-of-two scaling."""
+
+    params = {
+        "module__n_w_bits": n_bits,
+        "module__n_a_bits": n_bits,
+        "module__n_layers": 2,
+        "max_epochs": 10,
+    }
+
+    model = NeuralNetClassifier(**params)
+
+    x, y = load_data(NeuralNetClassifier)
+
+    model.fit(x, y)
+
+    adapter = PowerOfTwoScalingRoundPBSAdapter(model.quantized_module_)
+    round_pbs_patterns = adapter.process()
+    # Remove rounding in the network to perform inference without the optimization.
+    # We expect a network that was optimized with the power-of-two adapter
+    # to be exactly correct to the non-optimized one
+    for (_, node_op) in model.quantized_module_.quant_layers_dict.values():
+        if isinstance(node_op, QuantizedMixingOp):
+            node_op.rounding_threshold_bits = None
+            node_op.lsbs_to_remove = None
+
+    assert (
+        len(round_pbs_patterns) == 0
+    ), "Expected number of round PBS optimized patterns was not matched"
+    assert (
+        adapter.num_ignored_valid_patterns == 1
+    ), "Expected number of ignored round PBS optimizable patterns was not matched"
+
+    tlu_optimizer = TLUDeltaBasedOptimizer(
+        exactness=Exactness.APPROXIMATE,
+    )
+    cfg = Configuration(show_optimizer=False, additional_pre_processors=[tlu_optimizer])
+
+    # Compile the quantized model
+    model.compile(x, configuration=cfg, verbose=True)
+
+    # Find optimized TLUs
+    tlu_nodes = model.quantized_module_.fhe_circuit.graph.query_nodes(
+        custom_filter=lambda node: node.converted_to_table_lookup
+        and "opt_round_a" in node.properties["attributes"],
+        ordered=True,
+    )
+
+    # A single TLU should be present
+    assert len(tlu_nodes) == 1
+
+    count_reinterpret = 0
+    count_tlu = 0
+    for line in model.quantized_module_.fhe_circuit.mlir.split("\n"):
```

Comment:
should this be a function ? either in source (pytest) or in this file, but looks important